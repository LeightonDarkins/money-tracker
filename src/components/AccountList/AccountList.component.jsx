import React, { Component } from 'react'
import PropTypes from 'prop-types'
import Account from '../Account/Account.component.jsx'
import currency from '../../common/currency'
import './AccountList.scss'

class AccountList extends Component {
  constructor (props) {
    super(props)

    this.accountListElement = this.accountListElement.bind(this)
    this.accountBalanceAsCurrency = this.accountBalanceAsCurrency.bind(this)
  }

  componentDidMount () {
    this.props.fetchAccounts()
  }

  accountListElement () {
    if (this.props.isLoading) {
      return (<div className='account-list-spinner-container'>
        <i className='fa fa-circle-o-notch fa-spin fa-2x account-list-spinner' aria-hidden='true' />
      </div>)
    }

    if (this.props.accounts.length === 0) {
      return (<ul className='account-list'>
        <li className='placeholder'>No Accounts Available</li>
      </ul>)
    }

    return (<ul className='account-list'>
      {
        this.props.accounts.map(account => {
          return (<Account
            key={account.id}
            balance={account.balance}
            openingBalance={account.openingBalance}
            name={account.name}
            onAccountClick={() => this.props.onAccountClick(account.id)} />)
        })
      }
    </ul>)
  }

  accountBalanceAsCurrency (accounts) {
    let balance = accounts.reduce((a, b) => a + b.balance, 0)

    return currency.numberAsCurrency(balance)
  }

  render () {
    return (
      <div id='AccountList'>
        <section id='al-header' className='mt-header'>
          <h2>Accounts</h2>
          <div className='accounts-balance'>
            Balance: { this.accountBalanceAsCurrency(this.props.accounts) }
          </div>
        </section>
        <section className='mt-body scroll-view'>
          {this.accountListElement()}
        </section>
        <section className='mt-footer'>
          <button id='mt-add-transaction' className='account-list-action'
            onClick={this.props.onAddTransactionClick} >Add Transaction</button>
        </section>
      </div>
    )
  }
}

AccountList.propTypes = {
  accounts: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string.isRequired,
      openingBalance: PropTypes.number,
      name: PropTypes.string.isRequired
    })
  ).isRequired,
  onAccountClick: PropTypes.func.isRequired,
  onAddCategoryClick: PropTypes.func.isRequired,
  onAddTransactionClick: PropTypes.func.isRequired,
  onAddAccountClick: PropTypes.func.isRequired
}

export default AccountList
