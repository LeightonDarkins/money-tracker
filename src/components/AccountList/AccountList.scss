.account-list {
  list-style-type: none;
  padding: 0;
  margin: 0;
  min-height: 275px;
  max-height: 275px;
  overflow: auto;
}

.accounts-balance {
  text-align: center;
  font-size: 20px;
  width: 100%;
  margin-bottom: 20px;
}

.account-list-spinner-container {
  min-height: 275px;
  text-align: center;
}

.account-list-action-bar {
  position: absolute;
  bottom: 0px;
  display: flex;
  justify-content: center;
  width: 100%;
}

.account-list-action {
  margin: 5px 0 5px 0;
  width: 100%;
  color: #fff;
  background-color: #08B477;
  font-size: 16px;
  border: none;
}

.account-list-action:hover {
  color: #eee;
  cursor: pointer;
  background-color: #51EBB4;
}
