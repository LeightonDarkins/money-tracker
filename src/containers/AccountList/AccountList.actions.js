export const types = {
  accountClicked: 'ACCOUNT_CLICKED',
  fetchAccountsSucceeded: 'FETCH_ACCOUNTS_SUCCEEDED',
  fetchAccountsFailed: 'FETCH_ACCOUNTS_FAILED'
}

export const accountClicked = (id) => {
  return { type: types.accountClicked, id }
}

export const fetchAccountsSucceeded = (accounts) => {
  return { type: types.fetchAccountsSucceeded, accounts }
}
