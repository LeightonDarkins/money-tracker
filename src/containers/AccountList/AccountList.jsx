import React from 'react'
import axios from 'axios'
import _ from 'lodash'

import constants from '../../globals/constants.js'

import TestData from '../../globals/TestData.js'
import AccountListItem from '../../components/AccountListItem/AccountListItem.jsx'

class AccountList extends React.Component {
  constructor(props) {
    super(props)

    this.state = {
      accounts: []
    }
  }

  componentWillMount() {
    axios.get(`${constants.SERVER_URI}/accounts`)
    .then((data) =>{
      let accounts = data.data

      this.setState({ accounts: accounts })
    })
    .catch((error) => {
      console.error(error)
    })
  }

  accountsBalance() {
    const result = _.reduce(this.state.accounts, (sum, account) => {
      return sum + account.balance
    }, 0)

    return result;
  }

  render() {
    const accounts = this.state.accounts.map((account) => {
      return (
        <AccountListItem
          key={ account.id }
          name={ account.name }
          balance={ account.balance }/>
      )
    })

    if (this.state.accounts.length === 0) {
      return (
        <div>
          <h3>Accounts</h3>
          <p>No Accounts Found</p>
        </div>
      )
    } else {
      return (
        <div className='money-tracker-section'>
          <h3>Accounts</h3>
          <p>Balance: ${ this.accountsBalance() }</p>
          <ul className='list-group'>
            { accounts }
          </ul>
        </div>
      )
    }
  }
}

export default AccountList
