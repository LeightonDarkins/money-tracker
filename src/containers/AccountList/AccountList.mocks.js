module.exports = {
  accounts: [
    {
      id: 'account-1',
      name: 'first account',
      balance: 1,
      openingBalance: 0
    },
    {
      id: 'account-2',
      name: 'second account',
      balance: 2,
      openingBalance: 0
    }
  ]
}
