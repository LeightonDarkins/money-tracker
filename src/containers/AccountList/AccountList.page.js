module.exports = {
  getSpinner: (wrapper) => wrapper.find('.account-list-spinner'),
  getPlaceholder: (wrapper) => wrapper.find('.placeholder'),
  getAccountListItems: (wrapper) => wrapper.find('.account-list li')
}
