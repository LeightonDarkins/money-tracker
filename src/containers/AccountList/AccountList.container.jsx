import { connect } from 'react-redux'
import AccountList from '../../components/AccountList/AccountList.component.jsx'
import { accountClicked } from './AccountList.actions'

const mapStateToProps = state => {
  return {
    accounts: state.AccountList
  }
}

const mapDispatchToProps = dispatch => {
  return {
    onAccountClick: (id) => {
      dispatch(accountClicked(id))
    }
  }
}

const ConnectedAccountList = connect(
  mapStateToProps,
  mapDispatchToProps)(AccountList)

export default ConnectedAccountList
