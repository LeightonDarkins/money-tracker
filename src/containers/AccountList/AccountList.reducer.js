import { types } from './AccountList.actions'

const accounts = (state = [], action) => {
  switch (action.type) {
    case types.fetchAccountsSucceeded:
      return action.accounts
    default:
      return state
  }
}

export default accounts
