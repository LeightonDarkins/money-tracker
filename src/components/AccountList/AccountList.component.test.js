/* global describe, it, expect, jest, beforeEach, afterEach */

import React from 'react'
import { shallow } from 'enzyme'
import renderer from 'react-test-renderer'
import AccountList from './AccountList.component'
import th from '../../common/TestHelpers'

describe('AccountList', () => {
  describe('Snapshots', () => {
    it('should render with a spinner when loading', () => {
      const accountList = renderer.create(<AccountList
        accounts={[]}
        fetchAccounts={th.empty}
        onAccountClick={th.empty}
        onAddAccountClick={th.empty}
        onAddCategoryClick={th.empty}
        onAddTransactionClick={th.empty}
        isLoading />).toJSON()

      expect(accountList).toMatchSnapshot()
    })

    it('should render with a placeholder when no accounts are present', () => {
      const tree = renderer.create(<AccountList
        accounts={[]}
        fetchAccounts={th.empty}
        onAccountClick={th.empty}
        onAddAccountClick={th.empty}
        onAddCategoryClick={th.empty}
        onAddTransactionClick={th.empty} />).toJSON()

      expect(tree).toMatchSnapshot()
    })

    it('should render a list of accounts', () => {
      let testAccounts = [{
        id: 'testAccount1',
        name: 'testAccount1',
        balance: 500,
        openingBalance: 0
      }]

      const tree = renderer.create(
        <AccountList
          accounts={testAccounts}
          fetchAccounts={th.empty}
          onAccountClick={th.empty}
          onAddAccountClick={th.empty}
          onAddCategoryClick={th.empty}
          onAddTransactionClick={th.empty} />
      ).toJSON()

      expect(tree).toMatchSnapshot()
    })
  })

  describe('accountBalanceAsCurrency()', () => {
    const wrapper = shallow(<AccountList
      accounts={[]}
      onAccountClick={th.empty}
      onAddAccountClick={th.empty}
      onAddCategoryClick={th.empty}
      onAddTransactionClick={th.empty}
      />)
    const instance = wrapper.instance()

    it('returns $0.00 for empty array', () => {
      expect(instance.accountBalanceAsCurrency([])).toEqual('$0.00')
    })

    it('returns $10.00 for two accounts with balances of 500 cents', () => {
      let testAccounts = [{ balance: 500 }, { balance: 500 }]
      expect(instance.accountBalanceAsCurrency(testAccounts)).toEqual('$10.00')
    })

    it('returns $0.10 for two accounts with balances of 5 cents', () => {
      let testAccounts = [{ balance: 5 }, { balance: 5 }]
      expect(instance.accountBalanceAsCurrency(testAccounts)).toEqual('$0.10')
    })

    it('returns $1,000.00 for two accounts with balances of 50000', () => {
      let testAccounts = [{ balance: 50000 }, { balance: 50000 }]
      expect(instance.accountBalanceAsCurrency(testAccounts)).toEqual('$1,000.00')
    })
  })

  describe('Behaviour', () => {
    let wrapper
    let mockAddAccountClick, mockAddCategoryClick, mockAddTransactionClick

    beforeEach(() => {
      mockAddAccountClick = jest.fn()
      mockAddCategoryClick = jest.fn()
      mockAddTransactionClick = jest.fn()

      wrapper = shallow(<AccountList
        accounts={[]}
        onAccountClick={th.empty}
        onAddAccountClick={mockAddAccountClick}
        onAddCategoryClick={mockAddCategoryClick}
        onAddTransactionClick={mockAddTransactionClick} />)
    })

    afterEach(() => {
      mockAddAccountClick.mockReset()
      mockAddCategoryClick.mockReset()
      mockAddTransactionClick.mockReset()
    })

    it.skip('calls onAddAccountClick when the add account button is clicked', () => {
      wrapper.find('#mt-add-account').simulate('click')

      expect(mockAddAccountClick).toBeCalled()
    })

    it.skip('calls onAddCategoryClick when the add category button is clicked', () => {
      wrapper.find('#mt-add-category').simulate('click')

      expect(mockAddCategoryClick).toBeCalled()
    })

    it('calls onAddTransactionClick when the add transaction button is clicked', () => {
      wrapper.find('#mt-add-transaction').simulate('click')

      expect(mockAddTransactionClick).toBeCalled()
    })
  })
})
