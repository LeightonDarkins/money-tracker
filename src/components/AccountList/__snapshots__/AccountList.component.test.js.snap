exports[`AccountList Snapshots should render a list of accounts 1`] = `
<div
  id="AccountList">
  <section
    className="mt-header"
    id="al-header">
    <h2>
      Accounts
    </h2>
    <div
      className="accounts-balance">
      Balance: 
      $5.00
    </div>
  </section>
  <section
    className="mt-body scroll-view">
    <ul
      className="account-list">
      <li
        className="account-list-item"
        onClick={[Function]}>
        <div
          className="account-container">
          <span
            className="account-element account-name">
            testAccount1
          </span>
          <span
            className="account-element account-balance">
            $5.00
          </span>
          <div
            className="account-element account-opening-balance">
            Opening Balance: 
            $0.00
          </div>
        </div>
      </li>
    </ul>
  </section>
  <section
    className="mt-footer">
    <button
      className="account-list-action"
      id="mt-add-transaction"
      onClick={[Function]}>
      Add Transaction
    </button>
  </section>
</div>
`;

exports[`AccountList Snapshots should render with a placeholder when no accounts are present 1`] = `
<div
  id="AccountList">
  <section
    className="mt-header"
    id="al-header">
    <h2>
      Accounts
    </h2>
    <div
      className="accounts-balance">
      Balance: 
      $0.00
    </div>
  </section>
  <section
    className="mt-body scroll-view">
    <ul
      className="account-list">
      <li
        className="placeholder">
        No Accounts Available
      </li>
    </ul>
  </section>
  <section
    className="mt-footer">
    <button
      className="account-list-action"
      id="mt-add-transaction"
      onClick={[Function]}>
      Add Transaction
    </button>
  </section>
</div>
`;

exports[`AccountList Snapshots should render with a spinner when loading 1`] = `
<div
  id="AccountList">
  <section
    className="mt-header"
    id="al-header">
    <h2>
      Accounts
    </h2>
    <div
      className="accounts-balance">
      Balance: 
      $0.00
    </div>
  </section>
  <section
    className="mt-body scroll-view">
    <div
      className="account-list-spinner-container">
      <i
        aria-hidden="true"
        className="fa fa-circle-o-notch fa-spin fa-2x account-list-spinner" />
    </div>
  </section>
  <section
    className="mt-footer">
    <button
      className="account-list-action"
      id="mt-add-transaction"
      onClick={[Function]}>
      Add Transaction
    </button>
  </section>
</div>
`;
